lib/HttpClient.ts

```diff
@@ -297,48 +297,53 @@ export class HttpClient {
         info.options.method = method;
         info.options.headers = headers || {};
         info.options.headers["User-Agent"] = this.userAgent;
+        // if an agent is supplied in options use that agent for managing connections
+        if (!!this.requestOptions && !!this.requestOptions.agent) {
+            info.options.agent = this.requestOptions.agent;
```

Comment:
I think it's odd that if you supply an agent, the other options like proxy config and maxSockets are ignored.  That leaves a couple of options (1) layer those onto the agent supplied in specified in options (which means refactoring this flow in this function) or (2) simply add another option (or make it default) to re-use the agent out code creates.

The second option might be the best since it offers the performance benefit of re-using the socket without offloading the responsibilities of setting up the proxy etc... to the consumer.

Thoughts?

Reply:
2nd suggestion looks good to me. I have refactored the method to reuse agent across requests only if usePrivateAgent is set in the options. We also need to introduce a destroy function on HttpClient so that consumers can call it to close connections on private agent. We decided to share the agent across requests only if usePrivateAgent is true, since different agent might be needed for different get calls (e.g one request may bypass proxy and other might need to use tunnel agent). So only if consumers are sure that they want to use same agent across requests they can set usePrivateAgent to true. I have also added options of keepAlive and maxSockets to configure the agent.